zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,291 @@
+        let mut s: BitVec<Lsb0, u8> = BitVec::new();
+
+        // Prefix
+        s.append(&mut bitvec![1; 6]);
+
+        // Jubjub repr_J canonical byte encoding
+        // https://zips.z.cash/protocol/protocol.pdf#jubjub
+        let g_d_bytes = jubjub::AffinePoint::from(diversifier).to_bytes();
+        let pk_d_bytes = <[u8; 32]>::from(transmission_key);
+        let v_bytes = value.to_bytes();
+
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&g_d_bytes[..]));
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&pk_d_bytes[..]));
+        s.append(&mut BitVec::<Lsb0, u8>::from_slice(&v_bytes[..]));
+
+        let rcm = CommitmentRandomness(generate_trapdoor(csprng));
+
+        (
+            rcm,
+            NoteCommitment::from(windowed_pedersen_commitment(rcm.0, &s)),
+        )
+    }
+
+    /// Hash Extractor for Jubjub (?)
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concreteextractorjubjub
+    pub fn extract_u(&self) -> jubjub::Fq {
+        self.0.get_u()
+    }
+}
+
+/// A Homomorphic Pedersen commitment to the value of a note, used in Spend and
+/// Output Descriptions.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
+pub struct ValueCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
+
+impl fmt::Debug for ValueCommitment {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("ValueCommitment")
+            .field("u", &hex::encode(self.0.get_u().to_bytes()))
+            .field("v", &hex::encode(self.0.get_v().to_bytes()))
+            .finish()
+    }
+}
+
+impl From<[u8; 32]> for ValueCommitment {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(jubjub::AffinePoint::from_bytes(bytes).unwrap())
+    }
+}
+
+impl From<jubjub::ExtendedPoint> for ValueCommitment {
+    fn from(extended_point: jubjub::ExtendedPoint) -> Self {
+        Self(jubjub::AffinePoint::from(extended_point))
+    }
+}
+
+impl Eq for ValueCommitment {}
+
+impl From<ValueCommitment> for [u8; 32] {
+    fn from(cm: ValueCommitment) -> [u8; 32] {
+        cm.0.to_bytes()
+    }
+}
+
+/// LEBS2OSP256(repr_J(cv))
+///
+/// https://zips.z.cash/protocol/protocol.pdf#spendencoding
+/// https://zips.z.cash/protocol/protocol.pdf#jubjub
+impl ZcashSerialize for ValueCommitment {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&self.0.to_bytes())?;
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for ValueCommitment {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        Ok(Self(
+            jubjub::AffinePoint::from_bytes(reader.read_32_bytes()?).unwrap(),
+        ))
+    }
+}
+
+impl ValueCommitment {
+    /// Generate a new _ValueCommitment_.
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit
+    // TODO: accept an Amount instead?
```

Comment:
Might be safer?

Reply:
I think I'm missing something — it looks like this is taking the bytes of a scalar encoding a value, can that actually be replaced by an Amount?

Reply:
With enough `From` impls anything can be replaced with an Amount.

Reply:
I guess my actual goal is to turn the value into a scalar, and I assume is derived somehow from an `Amount` value somewhere above?

Reply:
So my guess is that yes, a `value` is expected to be turned into an element of the scalar field, @daira / @str4d

Reply:
Yes, there's a specification of ValueCommit in https://zips.z.cash/protocol/protocol.pdf#concretehomomorphiccommit:

![image](https://user-images.githubusercontent.com/44879/89464361-2dc70180-d725-11ea-8442-3ccccca84b79.png)

I'm not sure offhand about the `jubjub` API but the curve25519-dalek API provides `From` impls to construct scalars from machine-size words. I think it would be reasonable to provide 
```rust
impl<C> From<Amount<C>> for jubjub::Fr {
    fn from(a: Amount<C>) -> jubjub::Fr {
        if a.0 < 0 {
            jubjub::Fr::from(a.0.abs() as u64).neg()
        } else {
            jubjub::Fr::from(a.0)
        }
    }
}
```
This isn't quite optimal because it's not constant-time, but I don't think that that's essential to fix right now.

Reply:
I think this function could take an `Amount` as soon as that impl exists.

Reply:
Should this `impl` live with `Amount` or somewhere in here?

Reply:
Putting it with `Amount` seems like the best place.